Implement AttachmentCollectionResource

furinvader opened the pull request:
# Subset of #4768
## To-Do
- [x] Implement base version of the `AttachmentCollectionResource`
- [x] Make `attachments` property of a work package a `AttachmentCollectionResource`
- [x] Implement `opFileUpload` service
- [x] Move the functionality of the `wpAttachments` service to `AttachmentCollectionResource` 
  and `WorkPackageResource`
  - [x] Add `uploadAttachments` method to `WorkPackageResource`, that handles everything concerning the attachments after the upload
  - [x] Remove the `wpAttachments` service and its usages.
- [x] Fix file upload on work package creation
  - [x] Fix showing the uploaded file after work package creation
- [x] Fix specs
- [x] The attachments are displayed twice after uploading. This happens for newly created work packages as well as for editing existing ones. Upon reload, only one is displayed.
## Out of Scope / Dismissed
- ~~Refactor `halResourceTypes` service tests~~
- ~~Add override-child-type configuration to `halResourceTypes`~~

furinvader committed:
Move wpAttachments components

furinvader committed:
Export opTemplatesModule

furinvader committed:
Rename configModule to opConfigModule

furinvader committed:
Refactor wpAttachments directive

furinvader committed:
Refactor CollectionResource

furinvader committed:
Implement base version of the AttachmentCollectionResource

furinvader committed:
Refactor work package resource tests

furinvader committed:
Implement opFileUpload service

furinvader committed:
Implement upload method for AttachmentCollectionResource

furinvader committed:
Change the type of the opFileUpload upload result

furinvader committed:
Add uploadAttachments method to WorkPackageResource

furinvader committed:
Return updated resources in a promise

furinvader committed:
Fix uploadAttachments test

furinvader committed:
Refactor WorkPackageResource tests

furinvader committed:
Remove the wpAttachments service

furinvader committed:
Refactor the wpAttachments directive

ulferts commented on frontend/app/components/api/op-file-upload/op-file-upload.service.ts:
```diff
@@ -0,0 +1,66 @@
+
+export interface UploadResult {
+  uploads: IPromise<any>[];
+  upload: IPromise<any>;
```
I believe this should rather be called `uploaded` or `finished`

ulferts commented on frontend/app/components/wp-attachments/wp-attachments.directive.ts:
```diff
@@ -143,24 +147,23 @@ export class WorkPackageAttachmentsController {
 
   public remove(file):void {
-    if (!this.workPackage.isNew) {
-      if (file._type === 'Attachment') {
```
Thanks for merging those two statements 👍

ulferts commented:
File upload does not work upon work package creation. (added ToDo)

furinvader committed:
Make updateLinkedResources public and refactor the respective tests

furinvader committed:
Make halRequest return a rejected promise if no href is set

furinvader committed:
Reject update of linked resources, if they don't exist

furinvader committed:
Rename upload result 'uploads' to 'finished'

furinvader committed:
Fix upload of attachments for new work packages

furinvader committed:
Inform cache service of new WP only if the save was successful

furinvader committed:
Add first version of the wpAttachmentsUpload directive

furinvader committed:
Make AttachmentResource#upload upload pending attachments by default

furinvader committed:
Move directory files filter to the attachment collection resource

furinvader committed:
Add WorkPackageResource#canAddAttachments property

ulferts commented:
There are two bugs in here one of which was already present before:
1) present before - there is no upload notification displayed when an attachment is uploaded to a newly created work package
2) new - the attachment uploaded to a newly created work package is only displayed after one reloads or presses the "Show all attributes" button.

I think at least 2 should be fixed before merging, ideally also 1 if feasible.

furinvader committed:
Make use of uploadAttachments in upload directive

furinvader committed:
Remove pending attachments from attachment collection

furinvader committed:
Add uploadPendingAttachments method and use it

furinvader committed:
Separate the wpAttachmentsUpload directive

ulferts commented:
Todo about displaying uploaded filed upon wp creation added.

furinvader committed:
Implement wpAttachmentList directive

Replace the wpAttachments directive with the two newer ones.
Also add a new method to the WP resource.

furinvader committed:
Add updateElements method to CollectionResource

furinvader committed:
Update the attachment collection's elements where necessary

furinvader committed:
Remove non functional upload button

furinvader committed:
Remove # from delete attachment button href

furinvader commented:
Could not fix 1)
We should fix this, though, after the attachment button is done.

ulferts commented:
Now, the attachments are displayed twice after uploading. This happens for newly created work packages as well as for editing existing once. Upon reload, only one is displayed. 

Added ToDo.

furinvader commented:
Thanks for your feedback. It's fixed now.

furinvader committed:
Reset pending attachments after successful upload

ulferts commented:
Good to merge upon Travis' consent

ulferts commented:
Thanks for the work @furinvader

furinvader commented:
As for the first bug:
The notifications service automatically removes older notifications as soon as new ones appear.

I think this behaviour is faulty, because important information may be lost when multiple operations have to emit notifications.
